src/main/scala/com/fulcrumgenomics/bam/pileup/PileupBuilder.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam.pileup
+
+import com.fulcrumgenomics.bam.Bams
+import com.fulcrumgenomics.bam.api.SamRecord
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.fasta.SequenceDictionary
+import htsjdk.samtools.CigarOperator.{INSERTION => Insertion}
+
+import scala.collection.mutable.ArrayBuffer
+
+/** Class that provides methods to build and filter pileups.
+  *
+  * @param dict the sequence dictionary associated with the reference sequences.
+  * @param minBaseQ the minimum base quality that a base must have to contribute to a pileup.
+  * @param minMapQ the minimum mapping quality a read must have to contribute to a pileup.
+  * @param mappedPairsOnly if true, only allow read-pairs with both records mapped to contribute to a pileup.
+  * @param includeDuplicates if true, allow records flagged as duplicates to contribute to a pileup.
+  * @param includeSecondaryAlignments if true, allow records flagged as secondary alignments to contribute to a pileup.
+  * @param includeSupplementalAlignments if true, allow records flagged as supplementary alignments to contribute to a pileup.
+  * @param excludeMapPositionOutsideFrInsert if true, exclude any record of an FR pair where the site requested is outside the insert.
+  */
+class PileupBuilder(
```

Comment:
This is mostly identical to the one that previously existed in the codebase except it was profiled and its implementation sped up, especially with regard to record / pileup entry filtering.